routing/car_router.cpp

```diff
@@ -538,35 +495,95 @@ OsrmRouter::ResultCode OsrmRouter::CalculateRoute(m2::PointD const & startPoint,
+    return CarRouter::RouteNotFound;
   }
 }
 
-IRouter::ResultCode OsrmRouter::FindPhantomNodes(m2::PointD const & point,
-                                                 m2::PointD const & direction,
-                                                 TFeatureGraphNodeVec & res, size_t maxCount,
-                                                 TRoutingMappingPtr const & mapping)
+IRouter::ResultCode CarRouter::FindPhantomNodes(m2::PointD const & point,
+                                                m2::PointD const & direction,
+                                                TFeatureGraphNodeVec & res, size_t maxCount,
+                                                TRoutingMappingPtr const & mapping)
 {
   ASSERT(mapping, ());
-  helpers::Point2PhantomNode getter(*mapping, *m_pIndex, direction);
+  helpers::Point2PhantomNode getter(*mapping, m_index, direction);
   getter.SetPoint(point);
 
-  m_pIndex->ForEachInRectForMWM(getter, MercatorBounds::RectByCenterXYAndSizeInMeters(
-                                            point, kFeatureFindingRectSideRadiusMeters),
-                                scales::GetUpperScale(), mapping->GetMwmId());
+  m_index.ForEachInRectForMWM(getter, MercatorBounds::RectByCenterXYAndSizeInMeters(
+                                          point, kFeatureFindingRectSideRadiusMeters),
+                              scales::GetUpperScale(), mapping->GetMwmId());
 
   if (!getter.HasCandidates())
     return RouteNotFound;
 
   getter.MakeResult(res, maxCount);
   return NoError;
 }
+
+bool CarRouter::DoesEdgeIndexExist(Index::MwmId const & mwmId)
+{
+  MwmSet::MwmHandle const handle = m_index.GetMwmHandleById(mwmId);
+  if (!handle.IsAlive())
+  {
+    ASSERT(false, ("Mwm handle is not alive."));
+    return false;
+  }
+
+  MwmValue const * value = handle.GetValue<MwmValue>();
+  CHECK(value, ());
+  if (value->GetHeader().GetFormat() < version::Format::v8)
+    return false;
+
+  if (!value->m_cont.IsExist(EDGE_INDEX_FILE_TAG))
+    return false;
+
+  return true;
+}
+
+IRouter::ResultCode CarRouter::FindSingleRouteDispatcher(FeatureGraphNode const & source,
+                                                         FeatureGraphNode const & target,
+                                                         RouterDelegate const & delegate,
+                                                         TRoutingMappingPtr & mapping,
+                                                         Route & route)
+{
+  ASSERT_EQUAL(source.mwmId, target.mwmId, ());
+
+
+  IRouter::ResultCode result = IRouter::InternalError;
+  Route mwmRoute(GetName());
+
+  // @TODO It's not the best place for checking availability of edge index section in mwm.
+  // Probably it's better to keep if mwm has an edge index section in mwmId.
+  if (DoesEdgeIndexExist(source.mwmId))
+  {
+    if (!m_router)
+    {
+      LOG(LERROR, ("m_router is not initialized."));
+      return IRouter::InternalError;
+    }
+    // A* routing
```

Comment:
It's much better to get exact router name via some method, like m_router->GetName().